plsync/mlabconfig.py

```diff
@@ -526,6 +543,90 @@ def export_scraper_kubernetes_config(filename_template, experiments,
                     config_file.write(contents_tmpl.safe_substitute(config))
 
 
+def select_prometheus_experiment_targets(
+    experiments, select_regex, target_template, rsync_only):
+    """Selects and formats targets from experiments.
+
+    Args:
+      experiments: list of planetlab.Slice objects, used to enumerate hostnames.
+      select_regex: str, a regex used to choose a subset of hostnames. Ignored
+          if empty.
+      target_template: str, a template for formatting the target from the
+          hostname. e.g. {{hostname}}:7999, https://{{hostname}}/some/path
+      rsync_only: bool, skip experiments without rsync_modules.
+
+    Returns:
+      list of str, all selected and formatted targets.
+    """
+    targets = []
+    target_tmpl = BracketTemplate(target_template)
+    for experiment in experiments:
+        for _, node in experiment['network_list']:
+            # Skip experiments without an IP index.
+            if experiment['index'] is None:
+                continue
+
+            host = experiment.hostname(node)
+            # Consider all experiments or only those with rsync modules.
+            if not rsync_only or experiment['rsync_modules']:
+                if select_regex and not re.search(select_regex, host):
+                    continue
+                target = target_tmpl.safe_substitute({'hostname': host})
+                targets.append(target)
+
+    return targets
+
+
+def select_prometheus_node_targets(sites, select_regex, target_template):
+    """Selects and formats targets from site nodes.
+
+    Args:
+      sites: list of planetlab.Site objects, used to generate hostnames.
+      select_regex: str, a regex used to choose a subset of hostnames. Ignored
+          if empty.
+      target_template: str, a template for formatting the target from the
+          hostname. e.g. {{hostname}}:7999, https://{{hostname}}/some/path
+
+    Returns:
+      list of str, all selected and formatted targets.
+    """
+    targets = []
+    target_tmpl = BracketTemplate(target_template)
+    for site in sites:
+        for _, node in site['nodes'].iteritems():
+            if select_regex and not re.search(select_regex, node.hostname()):
+                continue
+            target = target_tmpl.safe_substitute({'hostname': node.hostname()})
+            targets.append(target)
+
+    return targets
+
+
+def targets_as_json(labels, targets):
+    """Generates a service discovery format for targets.
+
+    The service discovery targets are labeled with given labels.
+
+    Args:
+      labels: list of str, each element should be in the form: "key=value".
+      targets: list of str, the targets to scrape.
+
+    Returns:
+      str, the service discovery document as JSON.
+    """
+    labels = dict(filter(lambda x: len(x) == 2, [l.split('=') for l in labels]))
```

Comment:
Under what circumstances do you expect that labels might not be of the form "key=value"? If labels should always be of the form "key=value", should the script raise an error if one is passed that doesn't conform to this?